From 0282db555c3cf3fb2e47d2f696f1adbdad14f4cd Mon Sep 17 00:00:00 2001
From: Jan Pietek <jan.pietek@gmail.com>
Date: Thu, 5 Mar 2026 10:14:47 +0100
Subject: [PATCH 04/40] nvdec intra patch (legacy old tellyo code)

---
 libavcodec/cuviddec.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/libavcodec/cuviddec.c b/libavcodec/cuviddec.c
index 3fae9c12eb..7ad59f1740 100644
--- a/libavcodec/cuviddec.c
+++ b/libavcodec/cuviddec.c
@@ -342,6 +342,13 @@ static int CUDAAPI cuvid_handle_video_sequence(void *opaque, CUVIDEOFORMAT* form
     cuinfo.bitDepthMinus8 = format->bit_depth_luma_minus8;
     cuinfo.DeinterlaceMode = ctx->deint_mode_current;
 
+    if(avctx->gop_size == 100000) {
+       av_log(avctx, AV_LOG_INFO, "init decoder intra only\n");
+       cuinfo.ulIntraDecodeOnly = 1;
+    }
+
+    av_log(avctx, AV_LOG_INFO, "init decoder %d %d\n", ctx->nb_surfaces, avctx->refs);
+
     if (ctx->deint_mode_current != cudaVideoDeinterlaceMode_Weave && !ctx->drop_second_field)
         avctx->framerate = av_mul_q(avctx->framerate, (AVRational){2, 1});
 
-- 
2.53.0

