From 961ec93cbc352178ae5572c8d7f03f508edc3147 Mon Sep 17 00:00:00 2001
From: Jan Pietek <jan.pietek@gmail.com>
Date: Thu, 5 Mar 2026 12:25:44 +0100
Subject: [PATCH 10/20] v4l2 patch (legacy old tellyo code)

---
 libavdevice/v4l2.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavdevice/v4l2.c b/libavdevice/v4l2.c
index 0ae6872..644752f 100644
--- a/libavdevice/v4l2.c
+++ b/libavdevice/v4l2.c
@@ -584,7 +584,7 @@ static int mmap_read_frame(AVFormatContext *ctx, AVPacket *pkt)
         if (ctx->video_codec_id == AV_CODEC_ID_CPIA)
             s->frame_size = bytesused;
 
-        if (s->frame_size > 0 && bytesused != s->frame_size) {
+        if (s->frame_size > 0 && bytesused < s->frame_size) {
             av_log(ctx, AV_LOG_WARNING,
                    "Dequeued v4l2 buffer contains %d bytes, but %d were expected. Flags: 0x%08X.\n",
                    bytesused, s->frame_size, buf.flags);
-- 
2.53.0

