From ae1c26b083fbeb885b9850e1c54ec97ee9873d2a Mon Sep 17 00:00:00 2001
From: Jan Pietek <jan.pietek@gmail.com>
Date: Mon, 16 Mar 2026 14:47:10 +0100
Subject: [PATCH 15/20] crop cuda contd

---
 configure | 1 +
 1 file changed, 1 insertion(+)

diff --git a/configure b/configure
index a2e3e43..a9e3cc3 100755
--- a/configure
+++ b/configure
@@ -3314,6 +3314,7 @@ chromakey_cuda_filter_deps_any="cuda_nvcc cuda_llvm"
 colorspace_cuda_filter_deps="ffnvcodec"
 colorspace_cuda_filter_deps_any="cuda_nvcc cuda_llvm"
 crop_cuda_filter_deps="ffnvcodec"
+crop_cuda_filter_deps_any="cuda_nvcc cuda_llvm"
 hwupload_cuda_filter_deps="ffnvcodec"
 scale_npp_filter_deps="ffnvcodec libnpp"
 scale2ref_npp_filter_deps="ffnvcodec libnpp"
-- 
2.53.0

